integrations/jira/client.py

```diff
@@ -0,0 +1,242 @@
+                    )
+                    is_last = True
+                    continue
+
+                logger.error(
+                    f"HTTP error with status code: {e.response.status_code}"
+                    f" and response text: {e.response.text}"
+                )
+                raise
+            except httpx.HTTPError as e:
+                logger.error(f"HTTP occurred while fetching Jira data {e}")
+                raise
+        logger.info("Finished paginated request")
+        return
+
+    async def get_projects(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for projects in self._make_paginated_request(
+            f"{self.detail_base_url}/project/search"
+        ):
+            yield projects["values"]
+
+    async def _get_issues_from_board(
+        self, params: dict[str, str]
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for boards in self.get_all_boards():
+            for board in boards:
+                async for issues in self._make_paginated_request(
+                    f"{self.base_url}/board/{board['id']}/issue",
+                    params=params,
+                    is_last_function=lambda response: response["startAt"]
+                    + response["maxResults"]
+                    >= response["total"],
+                ):
+                    yield issues["issues"]
+
+    async def _get_issues_from_sprint(
+        self, params: dict[str, str]
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for sprints in self.get_all_sprints():
+            for sprint in sprints:
+                async for issues in self._make_paginated_request(
+                    f"{self.base_url}/sprint/{sprint['id']}/issue",
+                    params=params,
+                    is_last_function=lambda response: response["startAt"]
+                    + response["maxResults"]
+                    >= response["total"],
+                ):
+                    yield issues["issues"]
+
+    async def _get_issues_from_org(
+        self, params: dict[str, str]
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for issues in self._make_paginated_request(
+            f"{self.detail_base_url}/search",
+            params=params,
+            is_last_function=lambda response: response["startAt"]
+            + response["maxResults"]
+            >= response["total"],
+        ):
+            yield issues["issues"]
+
+    async def get_all_issues(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        config = typing.cast(JiraIssueResourceConfig, event.resource_config)
+        params = {}
+        if config.selector.jql:
+            params["jql"] = config.selector.jql
+            logger.info(f"Found JQL filter: {config.selector.jql}")
+
+        ISSUES_MAP = {
+            "board": self._get_issues_from_board,
+            "sprint": self._get_issues_from_sprint,
+            "all": self._get_issues_from_org,
```

Comment:
the client shouldn't be familiar with the event.resource_config, it should be as lean as possible and just expose an interface on how to interact with the third party, therefor let move the extraction of `event.resource_config` outside of the client.